state/azure/tablestorage/tablestorage.go

```diff
@@ -195,20 +193,29 @@ func (r *StateStore) writeRow(req *state.SetRequest) error {
 	}
 	entity.OdataEtag = etag
 
-	// InsertOrReplace does not support ETag concurrency, therefore we will try to use Update method first
-	// as it's more frequent, and then Insert
+	// InsertOrReplace does not support ETag concurrency, therefore we will use Insert to check for key existence
+	// and then use Update to update the key if it exists with the specified ETag
 
-	err := entity.Update(false, nil)
+	err := entity.Insert(storage.FullMetadata, nil)
 	if err != nil {
-		if isNotFoundError(err) {
-			// When entity is not found (set state first time) create it
-			entity.OdataEtag = ""
+		if etag == "" {
+			if req.Options.Concurrency == state.FirstWrite {
+				return state.NewETagError(state.ETagMismatch, err)
+			}
 
-			return entity.Insert(storage.FullMetadata, nil)
+			return entity.Update(true, nil)
 		}
+		err := entity.Update(false, nil)
+		if err != nil {
+			if isNotFoundError(err) {
+				return state.NewETagError(state.ETagMismatch, err)
+			}
+		}
+
+		return err
 	}
 
-	return err
+	return nil
```

Comment:
The logic LGTM, good job working it out. I would suggest heavily commenting it because the _why_ it works took me a while to unravel that from the structure of the code and the quirks of the conformance test expectations. 

I would also suggest explicitly checking for why Insert fails and handling expected cases with Update only.

```go
err := entity.Insert(storage.FullMetadata, nil)
if err != nil {
	// If Insert failed because item already exists, try to Update instead per Upsert semantics
	if isEntityAlreadyExistsError(err) {
		// Always Update using the etag when provided even if Concurrency != FirstWrite.
		// This behavior is enforced by test state.go:440 today, where we treat the presence of etag as trumping Concurrency.
		// In the future, we should error if an etag is provided when not using FirstWrite (see #2739)
		if etag != "" {
			uerr := entity.Update(false, nil)
			if uerr != nil {
				if isNotFoundError(uerr) {
					return state.NewETagError(state.ETagMismatch, uerr)
				}
				return uerr
			}
		} else if req.Options.Concurrency == state.FirstWrite {
			// Otherwise, if FirstWrite was set, but no etag was provided for an Update operation
			// explicitly flag it as an error. This is enforced by test state.go:541.
			// entity.Update itself does not flag the test case as a mismatch as it does not distinguish
			// between nil and "" etags, the initial etag will always be "", which would match on update.
			return state.NewETagError(state.ETagMismatch, errors.New("update with Concurrency.FirstWrite without ETag"))
		} else {
			// Finally, last write semantics without ETag should always perform a force update.
			return entity.Update(true, nil)
		}
	} else {
		// Any other unexpected error on Insert is propagated to the caller
		return err
	}
}

return nil
```

where 

```go
func isEntityAlreadyExistsError(err error) bool {
	azureError, ok := err.(storage.AzureStorageServiceError)

	return ok && azureError.Code == "EntityAlreadyExists"
}
```